### PR TITLE
Argent account implementation

### PR DESCRIPTION
**Pull request by dribeiro-ShardLabs**
## Usage related changes
- Argent Account implementation can now be used:
  - additional functionalities for guardian setting `await loadedAccount.setGuardian(account.guardianPrivateKey)` ;
  - can be deployed and used the same way as the OpenZeppelin implementation, e.g `account = await starknet.deployAccount("ArgentAccount")`, but to load from an address or to change the guardian, must import `import { ArgentAccount } from "@shardlabs/starknet-hardhat-plugin/dist/account"
` and cast as ArgentAccount `const loadedAccount = (await starknet.getAccountFromAddress(accountAddress, privateKey, "ArgentAccount")) as ArgentAccount`
- Both Argent and OZ implementations now support multicall/invoke, described in README

## Development related changes
- Implemented Argent Account, changed way OZ implementation invokes, now it uses multicall/invoke, without breaking changes

# Checklist:

- [x] I have formatted the code
- [x] I have performed a self-review of the code
- [x] I have rebased to the base branch
- [x] I have documented the changes
- [x] I have updated the tests
- [x] I have created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/Shard-Labs/starknet-hardhat-example).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,52 +3,64 @@
 If you've used Hardhat 👷‍♀️👷‍♂️ and want to develop for Starknet <img src="https://starkware.co/wp-content/uploads/2021/07/Group-177.svg" alt="starknet" width="18"/>, this plugin might come in hand. If you've never set up a Hardhat project, check out [this guide](https://hardhat.org/tutorial/creating-a-new-hardhat-project.html).
 
 ## Contents
-- [Install](#install)
```
Comment (FabijanC) [inline README.md:6]: Similar to config.yml, I see formatting has been applied. I've been avoiding that change as it makes the review difficult. If you could revert the formatting changes, then I'd be able to review more easily.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: .circleci/config.yml [hunk 2]
```diff
@@ -4,14 +4,14 @@ version: 2.1
 # Executor name definition
 ##########################
 executors:
-  linux-node14:
```
[the remark this answers is not in the document]
Comment (dribeiro-ShardLabs) [reply .circleci/config.yml:7]: Yeah, the update for the circleci cache regarding cairo version

File: src/type-extensions.ts [hunk 3]
```diff
@@ -147,7 +147,7 @@ declare module "hardhat/types/runtime" {
              * @param accountType the enumerator value of the type of Account to use
              * @returns an Account object
              */
-            deployAccount: (accountType: AccountImplementationType) => Promise<Account>;
```
[the remark this answers is not in the document]
Comment (dribeiro-ShardLabs) [reply src/type-extensions.ts:150]: Leftover attempt to expose the class, will revert

File: src/account-utils.ts [hunk 4]
```diff
@@ -139,3 +191,27 @@ async function assertArtifact(
         fs.writeFileSync(path.join(artifactsTargetPath, artifact), response.data);
     }
 }
+
+export function parseMulticallOutput(
```
[the remark this answers is not in the document]
Comment (dribeiro-ShardLabs) [reply src/account-utils.ts:195]: You mean before we changed to use adapt output directly? this is different, it does the adapt output for the response array of the multicall, parsed according to the function abi

File: src/account.ts [hunk 5]
```diff
@@ -176,4 +168,253 @@ export class OpenZeppelinAccount extends Account {
...
+     * Performes multiple invokes as a single transaction through this account
+     * @param callParameters an array with the paramaters for each invoke
+     * @returns the transaction hash of the invoke
+     */
+    async multiInvoke(callParameters: CallParameters[]): Promise<string> {
+        // Invoke only returns one transaction hash, as the multiple invokes are done by the account contract, but only one is sent to it.
+        return (await this.multiInvokeOrMultiCall("invoke", callParameters)).toString();
+    }
+
+    private async multiInvokeOrMultiCall(choice: Choice, callParameters: CallParameters[]) {
```
[the remark this answers is not in the document]
Comment (dribeiro-ShardLabs) [reply src/account.ts:312]: `multiInvokeOrMultiCall` is different for each, Argent for example has the guardian signatures, which do not exist in OZ account

File: src/types.ts [hunk 6]
```diff
@@ -373,6 +373,15 @@ export class StarknetContract {
         this.feederGatewayUrl = config.feederGatewayUrl;
     }
 
+    get address(): string {
+        return this._address;
+    }
+
+    set address(address: string) {
```
[the remark this answers is not in the document]
Comment (dribeiro-ShardLabs) [reply src/types.ts:380]: Address for StarknetContract is set when deploying and getting contract from an address (lines 349 and  301)

**Review by FabijanC — APPROVED** [review]

**Pull request merged**